src/crypto/keccak.c

```diff
@@ -212,3 +212,51 @@ void keccak_finish(KECCAK_CTX * ctx, uint8_t *md){
         memcpy_swap64le(md, ctx->hash, KECCAK_DIGESTSIZE / sizeof(uint64_t));
     }
 }
+
+void hmac_keccak_init(hmac_keccak_state *S, const uint8_t *_key, uint64_t keylen) {
+    const uint8_t *key = _key;
+    uint8_t keyhash[HASH_SIZE];
+    uint8_t pad[KECCAK_BLOCKLEN];
+    uint64_t i;
+
+    if (keylen > KECCAK_BLOCKLEN) {
+        keccak(key, keylen, keyhash, HASH_SIZE);
+        key = keyhash;
+        keylen = HASH_SIZE;
+    }
+
+    keccak_init(&S->inner);
+    memset(pad, 0x36, KECCAK_BLOCKLEN);
+    for (i = 0; i < keylen; ++i) {
+        pad[i] ^= key[i];
+    }
```

Comment:
This allows keys with size < 32. That loop looks like a two byte key of 0x42 0x00 would behave the same as a one byte key of 0x42. Is that intended ?

Reply:
I think this is according to the RFC 2104.

```
B = hash function block length
ipad = the byte 0x36 repeated B times
opad = the byte 0x5C repeated B times.
HMAC(K, text) = H(K XOR opad, H(K XOR ipad, text))
```

Also according to the section 3:
"The key for HMAC can be of any length (keys longer than B bytes are first hashed using H). 
 However, less than L (size of hash output) bytes is strongly discouraged as it would decrease the security strength of the function. "